softwareDesign.md

```diff
@@ -0,0 +1,3 @@
+﻿:computer:  **Software Design** :floppy_disk:  
+
+We focused on Python this week.  It is a high level language that is powerful, straight forward, and well constructed.  It is an object oriented programming language, meaning everything including functions are basically objects. One particular design pattern that I like so far is how Python favors object composition over inheritance.  I like it because the result is I end up creating fewer classes or subclasses as opposed to wrapping one class in another class.  This would be useful in situations where we want to restrict the methods of the wrapped class to expose.  This means the same construct can be used on a database during one day, and on another day it can be whatever else we wish. Regardless of what language I'm coding in or learning, one common anti-pattern that I encounter frequently is Readability.  The more simple and readable that code is, the smaller the chances are that it contains bugs.  It is also less likely to be as efficient as it could be.  One way I address this issue is through refactoring,  while keeping in mind the benefits or sacrifices of each step in refactoring.  I try to avoid unecessary nested loops, unused variables, I try to have meaningful names for both functions and variables.  And lastly, I try to make sure that the functions being performed are done so in an efficient manner- and an example of that would be with for loops.  It is crucial to see if we are integrating at a high level since these have an effect on performance.  Instead of doing something like `for x in range(len(array)):` we could do something like `if x in array:` and we would avoid the for loop altogether.
```

Comment:
It's definitely good practice to make an effort to not carelessly name variables, functions,  etc., especially as your application starts to get larger and more complicated. Also, avoiding unnecessary iterations in your code will keep you from taking too much time to do simple tasks.